components/datetime/src/pattern/mod.rs

```diff
@@ -107,8 +110,174 @@ impl From<Vec<PatternItem>> for Pattern {
+                    let mut needs_escaping = false;
+                    for ch in literal.chars() {
+                        if ch.is_ascii_alphabetic() || ch == '\'' {
+                            needs_escaping = true;
+                            break;
+                        }
+                    }
+
+                    if needs_escaping {
+                        let mut ch_iter = literal.trim_end().chars().peekable();
+
+                        // Do not escape the leading whitespace.
+                        while let Some(ch) = ch_iter.peek() {
+                            if ch.is_whitespace() {
+                                string.push(*ch);
+                                ch_iter.next();
+                            } else {
+                                break;
+                            }
+                        }
+
+                        // Wrap in "'" and escape "'".
+                        string.push('\'');
+                        for ch in ch_iter {
+                            if ch == '\'' {
+                                // Escape a single quote.
+                                string.push('\\');
+                            }
+                            string.push(ch);
+                        }
+                        string.push('\'');
+
+                        // Add the trailing whitespace
+                        for ch in literal.chars().rev() {
+                            if ch.is_whitespace() {
+                                string.push(ch);
+                            } else {
+                                break;
+                            }
+                        }
+                    } else {
+                        string.push_str(literal);
+                    }
+                }
+            }
+        }
+
+        string
+    }
+}
+
 impl FromIterator<PatternItem> for Pattern {
     fn from_iter<I: IntoIterator<Item = PatternItem>>(iter: I) -> Self {
         Self::from(iter.into_iter().collect::<Vec<_>>())
     }
 }
+
+#[cfg(feature = "provider_serde")]
+struct DeserializePatternVisitor;
+
+#[cfg(feature = "provider_serde")]
+impl<'de> de::Visitor<'de> for DeserializePatternVisitor {
+    type Value = Pattern;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "Expected to find a valid pattern.")
+    }
+
+    fn visit_str<E>(self, pattern_string: &str) -> Result<Self::Value, E>
+    where
+        E: de::Error,
+    {
+        // Parse a string into a list of fields.
+        let pattern = match Pattern::from_bytes(pattern_string) {
+            Ok(p) => p,
+            Err(err) => {
+                return Err(E::custom(format!(
+                    "The pattern \"{}\" could not be parsed: {:?}",
+                    pattern_string, err
+                )));
+            }
+        };
+        Ok(pattern)
+    }
+}
+
+#[cfg(feature = "provider_serde")]
+impl<'de> Deserialize<'de> for Pattern {
```

Comment:
Okay, so here's the thing.

The CLDR date pattern syntax is almost definitely *not* the best way to store patterns in the data provider.  I've been saying this for a while, including in my latest update to data_pipeline.md.  It is designed for human readability, not for performance.  What you're doing here is to make the CLDR date pattern syntax the de-facto standard for the ICU4X data model.  Adding custom serde comes with costs; we should only do it if we're very confident about the data model.

Why can't you just store the parsed array of pattern tokens in the data provider?

Reply:
Main issue: #523 

We decided to keep the exotic type in the data struct.